atlascope/core/fields/cube.py

```diff
@@ -0,0 +1,56 @@
+from collections import UserList
+import re
+from typing import Iterable
+
+from django.db.models import Field, FloatField, Func
+from psycopg2.extensions import AsIs
+
+from atlascope.core.utils import register_adapter
+
+__all__ = ['Cube', 'CubeDistance', 'CubeField']
+
+
+class Cube(UserList):
+    data: Iterable[float]
+
+
+@register_adapter(Cube)
+def adapt_cube(cube: Iterable[float]) -> str:
+    quoted_vector = [str(x) for x in cube]
+    return AsIs("cube(array[%s]::float[])" % ",".join(quoted_vector))
+
+
+def parse_cube(cube_string: str) -> Cube:
+    quoted_vector = re.findall(r'-?\d+(?:\.\d+)?(?:[eE]-?\d+)?', cube_string)
+    return Cube([float(s) for s in quoted_vector])
+
+
+class CubeField(Field):
+
+    description = "An n-d vector using the PostgreSQL cube module"
+
+    def db_type(self, connection):
+        return 'cube'
+
+    def from_db_value(self, value, expression, connection):
+        if isinstance(value, Cube) or value is None:
+            return value
+        return parse_cube(value)
+
+    def to_python(self, value):
+        if isinstance(value, Cube) or value is None:
+            return value
+        return parse_cube(value)
+
+
+class CubeDistance(Func):
```

Comment:
Out of curiosity: what's an example of how you would end up using this?

Reply:
The `CubeDistance` is how we're finding similar nuclei. See [this line](https://github.com/atlascope/atlascope/pull/191/files#diff-b6cbc5f32a7e320716eb17b588e7207a7f5a071645c18ec9f2696051c9e5ff02R126). Also, see the [relevant PostgreSQL docs](https://www.postgresql.org/docs/current/cube.html#CUBE-OPERATORS-TABLE). 

Two nuclei that are similar will have a small Euclidean distance between them since each nucleus is indexed as a 100D vector with unit length. The GiST index on the `CubeField` clusters nearby nuclei on a single page. So, ordering all nuclei by their distance to a target nucleus and picking just the top few results in an $\mathcal{O}(1)$ operation.